registry/storage/ocimanifesthandler.go

```diff
@@ -107,12 +116,13 @@ func (ms *ocischemaManifestHandler) verifyManifest(ctx context.Context, mnfst oc
 				err = distribution.ErrBlobUnknown // just coerce to unknown.
 			}
 
+			if err != nil {
+				dcontext.GetLogger(ms.ctx).WithError(err).Debugf("failed to ensure exists of %v in manifest service", descriptor.Digest)
+			}
 			fallthrough // double check the blob store.
 		default:
-			// forward all else to blob storage
-			if len(descriptor.URLs) == 0 {
-				_, err = blobsService.Stat(ctx, descriptor.Digest)
-			}
+			// check the presence
+			_, err = blobsService.Stat(ctx, descriptor.Digest)
```

Comment:
Why did you remove the check for the optional URLs? I understand this switch clause is either hit via `fallthrough` or if "non-image" `mediaType` is processed.

Reply:
The config media-type can hit the default branch. If I understand correctly, for the distributable blob, the registry should check the presence whether it has optional URLs or not. :)

Does it make senses to you? 😄

Reply:
> The config media-type can hit the default branch

Correct. So can `index`.

> If I understand correctly, for the distributable blob, the registry should check the presence whether it has optional URLs or not

Semantically this makes sense to me, but I can't seem to find any mention of this in the spec. I might be missing something, though.

Reply:
Yeah. The spec just mentions for non-distributable layer https://github.com/opencontainers/image-spec/blob/v1.0.1/layer.md#non-distributable-layers. And I think the spec is based on `content-addressable`.

```
FROM: https://github.com/opencontainers/image-spec/blob/859973e32ccae7b7fc76b40b762c9fff6e912f9e/manifest.md

There are three main goals of the Image Manifest Specification. The first goal is content-addressable images, by supporting an image model where the image's configuration can be hashed to generate a unique ID for the image and its components.
```

So, for the config/manifest blob, the digest is the index for download.

Reply:
I'd like more people to weigh in on this. Your approach makes sense to me, but I wonder why has the optional URL check lasted for so long without anyone shouting.

Reply:
@milosgajdos the URLs are not widely used and recently even been the source of a client CVE. I don't think there are legitimately use cases that would have been hitting this condition and have I seen non-standard media types which make use of URLs at all. If someone does shout, I'll love to see what hacky thing they are doing :)

Reply:
@milosgajdos and @dmcgowan thanks!